### ISSUE TITLE
Questions about dataflow and multi process runner

### ISSUE THREAD

**Issue by Rov67777** [tensorpack/tensorpack#1258]
Hello,

my problem is the following, my data are very large images (3D images that can take up to 1Go) therefore the bottleneck for the dataflow efficiency is mostly related to hard drive reading time performance when loading the images on the memory. I have read the documentation about dataflow and how to make them more efficient (here https://tensorpack.readthedocs.io/tutorial/efficient-dataflow.html) but I still encounter difficulties to understand some of tensorpack functions.

1. Is there any use to call dataflow.MultiProcessRunner with num_proc=1 ? The documentation is suggesting to do so but I don't quite understand the rationale if there is not actually multi processor running ?

2. will dataflow.MultiProcessRunner with num_proc >2 actually have any impact on the efficiency if the bottleneck is disk read performance ? My feeling would be that adding more processor to do the job won't have any impact if the hard drive reading capacity is already at maximum.

3. what is ZeroMQ for communication used in dataflow.MultiProcessRunnerZMQ and the difference with the standard dataflow.MultiProcessRunner  ? 

What solution would you recommend in my case ?

**Comment by ppwwyyxx** [comment]
> Is there any use to call dataflow.MultiProcessRunner with num_proc=1 ?

Yes. It serves as prefetching & latency hiding. The dataflow runs in a separate process in parallel.

> will dataflow.MultiProcessRunner with num_proc >2 actually have any impact on the efficiency if the bottleneck is disk read performance ? 

It will not.

> what is ZeroMQ for communication used in dataflow.MultiProcessRunnerZMQ and the difference with the standard dataflow.MultiProcessRunner ?

http://zeromq.org/  It's faster than MultiProcessRunner.

You are recommended to benchmark first to quantify where is the bottleneck.

**Comment by phizaz** [comment]
Is it possible that `dataflow.MultiProcessRunner` with `num_proc=1` be a bottleneck itself? By removing it, I could reach better results. 
However, there comes a problem that `LocallyShuffleData` seems to memory leak.

**Comment by ppwwyyxx** [comment]
> Is it possible that dataflow.MultiProcessRunner with num_proc=1 be a bottleneck itself? By removing it, I could reach better results.

It is possible. Because it needs to send data between the processes it could be bottleneck if the communication alone is more expensive than getting the data.

> there comes a problem that LocallyShuffleData seems to memory leak.

Could you reproduce the issue with simpler code and post it?

**Comment by phizaz** [comment]
@ppwwyyxx Okay, it is not a memory leak. It is just that without `MultiProcessRunner` the whole data flow got forked by many subsequent `MultiProcessRunnerZMQ`'s such that each of them has their own `LocallyShuffleData`. This proves to be too large for my RAM effectively I see a spike in memory usage.

This begs a question isn't there a way to prevent forking the flow without using `MultiProcessRunner`?

**Comment by Rov67777** [comment]
My understanding is that if you apply the LocallyShuffleData after the MultiProcessRunnerZMQ with num_proc=K it should only has one LocallyShuffleData that gets feed by K forked process of MultiProcessRunnerZMQ?

**Comment by phizaz** [comment]
@Rov67777 It seems reasonable to think so, but it doesn't work I just tested. I think there needs to be a different kind of flow that some how "joins".

**Comment by phizaz** [comment]
By the way removing `MultiProcessRunner` seems unreasonable as suggested in 😢: https://tensorpack.readthedocs.io/tutorial/efficient-dataflow.html

> Since we are reading the database sequentially, having multiple forked instances of the
base LMDB reader will result in biased data distribution. Therefore we use MultiProcessRunner to
launch the base DataFlow in only one process, ...

**Comment by ppwwyyxx** [comment]
> It is just that without MultiProcessRunner the whole data flow got forked by many subsequent MultiProcessRunnerZMQ's such that each of them has their own LocallyShuffleData

I did not know you're using them together so I was not able to comment anything on that. For the same reason, please ask questions with details: please ask questions with what you did and what you observed. Please include what you did and what you observed by pasting code & logs/screenshots, not by words. The [issue template](https://github.com/tensorpack/tensorpack/blob/master/.github/ISSUE_TEMPLATE/unexpected-problems---bugs.md) will help you include necessary information.
If you have trouble improving the performance, please include details following the [performance guide](https://tensorpack.readthedocs.io/tutorial/performance-tuning.html)

**Comment by ppwwyyxx** [comment]
Feel free to post a new issue with details if you still have questions to ask.
